### PR TITLE
Testing sync and async in one test.

### PR DESCRIPTION
**Pull request by kasobol-msft**
This introduces:
- test utilities for sync-async tests authoring
- refactored test hacked so far

TODO:
- it doesn't work well with other JUnit templates... but mixing JUnit templates doesn't work at all, i.e. can't mix RepeatedTest with ParametrizedTest... Either JUnit is going to fill this gap some day (there seems to be plenty of asks for that) or we need to invent some workaround until that happens.

### CODE REVIEW

**Comment by pallavit** [comment]
> API changes have been detected in `com.azure:azure-core`. You can review API changes [here](https://apiview.dev/Assemblies/Review/36605e430fb64abc9c0dec8d2be701aa?diffOnly=True&diffRevisionId=4925f9cecf14443ab60d5edbdc340292)


Also this always confusing me. Do you know why do we diff against the latest available version vs the latest available GA version?

**Review by alzimmermsft — APPROVED** [review]

File: sdk/core/azure-core-test-junit-extensions/README.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  <version>1.0.0-beta.1</version>
+</dependency>
+```
+[//]: # ({x-version-update-end})
+
+## Key concepts
+
+
+## Examples
+
```
Comment (samvaity) [inline sdk/core/azure-core-test-junit-extensions/README.md:23]: Will this be added in a follow-up PR?

Comment (kasobol-msft) [reply sdk/core/azure-core-test-junit-extensions/README.md:23]: Yeah, I left TODO in sdk/core/azure-core-test-junit-extensions/src/samples/java/com/azure/core/test/junit/extensions/ReadmeSamples.java

File: sdk/core/azure-core-test-junit-extensions/src/test/java/com/azure/core/test/junit/extensions/SyncAsyncTest.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
+public class SyncAsyncTest {
+
+    private static final AtomicInteger SYNC_INVOCATIONS = new AtomicInteger();
+    private static final AtomicInteger ASYNC_INVOCATIONS = new AtomicInteger();
+
+    @Order(1)
+    @com.azure.core.test.junit.extensions.annotation.SyncAsyncTest
```
Comment (samvaity) [inline sdk/core/azure-core-test-junit-extensions/src/test/java/com/azure/core/test/junit/extensions/SyncAsyncTest.java:22]: any reason for the fully qualified namespace?

**Review by samvaity — APPROVED** [review]

**Pull request merged**